libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -173,21 +173,10 @@ namespace eosio { namespace chain {
                trx_context.resume_billing_timer();
             });
             trx_context.pause_billing_timer();
-            IR::Module module;
             std::vector<U8> bytes = {
                 (const U8*)codeobject->code.data(),
                 (const U8*)codeobject->code.data() + codeobject->code.size()};
-            try {
-               Serialization::MemoryInputStream stream((const U8*)bytes.data(),
-                                                       bytes.size());
-               WASM::scoped_skip_checks no_check;
-               WASM::serialize(stream, module);
-               module.userSections.clear();
-            } catch (const Serialization::FatalSerializationException& e) {
-               EOS_ASSERT(false, wasm_serialization_error, e.message.c_str());
-            } catch (const IR::ValidationException& e) {
-               EOS_ASSERT(false, wasm_serialization_error, e.message.c_str());
-            }
+            IR::Module module = trx_context.control.get_wasm_module_cache().get_module(code_hash, vm_type, vm_version, bytes);
 
             wasm_instantiation_cache.modify(it, [&](auto& c) {
```

Comment:
Is there a reason not to have a shared `wasm_instantiation_cache` ?

Reply:
Seems like they could use the same boost multi index that stores both the `IR::Module` and the `wasm_instantiated_module_interface`.

Reply:
I initially tried shared `wasm_instantiation_cache`. It did not work, likely due to some other deep complications. For RCs, I did not want to make too tricky changes.

Reply:
What problems did you run into with a shared `wasm_instantiation_cache`?

Reply:
Instantiated modules rely on runtime interface and cannot be simply shared. nodeos can run a few multi-threaded read-only trx for a few seconds and crashes with memory corruption. I plan to look into this further after RC. For now focus on cleaning up other loose ends.